src/module_wrap.cc

```diff
@@ -0,0 +1,531 @@
+#include <algorithm>
+#include <limits.h>  // PATH_MAX
+#include <sys/stat.h>  // S_IFDIR
+#include "module_wrap.h"
+
+#include "env.h"
+#include "node_url.h"
+#include "util.h"
+#include "util-inl.h"
+
+namespace node {
+namespace loader {
+
+using node::url::URL;
+using node::url::URL_FLAGS_FAILED;
+using v8::Context;
+using v8::EscapableHandleScope;
+using v8::Function;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::Integer;
+using v8::IntegrityLevel;
+using v8::Isolate;
+using v8::JSON;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::Module;
+using v8::Object;
+using v8::Persistent;
+using v8::Promise;
+using v8::ScriptCompiler;
+using v8::ScriptOrigin;
+using v8::String;
+using v8::Value;
+
+static const char* EXTENSIONS[] = {".mjs", ".js", ".json", ".node"};
+std::map<int, std::vector<ModuleWrap*>*> ModuleWrap::module_map_;
+
+ModuleWrap::ModuleWrap(Environment* env,
+                       Local<Object> object,
+                       Local<Module> module,
+                       Local<String> url) : BaseObject(env, object) {
+  Isolate* iso = Isolate::GetCurrent();
+  module_.Reset(iso, module);
+  url_.Reset(iso, url);
+}
+
+ModuleWrap::~ModuleWrap() {
+  Local<Module> module = module_.Get(Isolate::GetCurrent());
+  std::vector<ModuleWrap*>* same_hash = module_map_[module->GetIdentityHash()];
+  auto it = std::find(same_hash->begin(), same_hash->end(), this);
+
+  if (it != same_hash->end()) {
+    same_hash->erase(it);
+  }
+
+  module_.Reset();
+}
+
+void ModuleWrap::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  Isolate* iso = args.GetIsolate();
+
+  if (!args.IsConstructCall()) {
+    env->ThrowError("constructor must be called using new");
```

Comment:
If `ModuleWrap` is not intended to be public API, I'd prefer these to be `CHECK`s